### ISSUE TITLE
Refactor protocol definitions into separate library

### ISSUE THREAD

**Issue by pmconrad** [bitshares/bitshares-core#1506]
**User Story**
As a software architect I want to enforce a clear separation of the BitShares protocol definition and the BitShares Core implementation.

The `libraries/chain` directory currently contains a `protocol` subdirectory, which forms a weak architectural boundary. This is also reflected in a separate `protocol` subdirectory for the relevant include files. It is an unwritten (?) rule that code in the `protocol` directory must not depend on code or definitions from the upper levels. This rule can easily be violated. It should therefore be enforced by moving the protocol-specific files into a separate library of their own.

(Steem did the same a long time ago.)

**Impacts**
- [x] Build (the build process or something prior to compiled code)
- [x] Other (please add below)
Software architecture

## CORE TEAM TASK LIST
- [x] Evaluate / Prioritize Feature Request
- [ ] Refine User Stories / Requirements
- [ ] Define Test Cases
- [ ] Design / Develop Solution
  - [x] Assigned: @nathanhourt 
  - [ ] Estimation: Pending
  - [ ] Break FC out from a submodule to an external dependency which is expected to be installed in the system
    - [ ] `protocol`
    - [ ] `db`
- [ ] Perform QA/Testing
- [ ] Update Documentation

**Comment by oxarbitrage** [comment]
I agree with this. It is legal right now to include for example an object definition into a protocol header file for whatever reason. I only realized this was not a good idea after seeing this was not done anywhere else.

**Comment by jmjatlanta** [comment]
Of course it is possible, I've (unwittingly) done it. Fortunately, it didn't pass code review. This simple change helps developers see the lines between components.

**Comment by abitmore** [comment]
In principle I agree that this is a good move. If this is done, the only downside I can think of, is it's harder to track old changes done on the relocated files.

**Comment by ryanRfox** [comment]
I too support any effort to refactor our code to be more readable and reinforce best practices.  I notice this received the `small` label. I'm wondering if it may also qualify as `good first issue`. I realize it's `low` priority, but if you feel it can help us on-board new contributors now and make it easier for others to use our code going forward, I'd appreciate identifying a Community Claims contributor(s).

**Comment by pmconrad** [comment]
I think it requires more knowledge of cmake and build processes than of the internals of our codebase. Added the "good first issue" label.

**Comment by nathanhourt** [comment]
I'll take this issue, if no one else already wants it. I think it shouldn't be too difficult, but I'm going to explore some other changes simultaneously that I think are worth doing, most notably:

 - Break FC out from a submodule to an external dependency which is expected to be installed in the system
 - Use the Meson build system for the protocol library, and evaluate the benefits/viability of converting the larger Bitshares repository to Meson from CMake

As we work to further modularize Bitshares, I envision a proliferation of independently building subprojects, which most or all depend on FC. I would like to see these projects live in separate repos with independent build systems, as placing them all in a single repo with a single build system allows the pieces to interact in less well-defined ways, which somewhat defeats the purpose of modularization. We may still have a parent project that connects all the subprojects together for a single build, but the subprojects shouldn't require this.

Having FC as a submodule of each of these subprojects is clearly undesirable, and having all of the subprojects depend on an overarching parent project/build which has an FC submodule is, again, purpose defeating.

The above discussion takes a somewhat mid- to long-term view, but in the shorter term, we want to break out the protocol library, which requires FC, as too will the main bitshares repo. So treating FC as an external dependency becomes relevant now and paves the road for further modularization later on.

As to the conversion from CMake to Meson, perhaps I will initially create both build systems for the protocol library to allow for a direct side-by-side comparison at small scale, at which point I predict the advantages of Meson to be apparent.

Thoughts?

**Comment by pmconrad** [comment]
IMO keeping FC as a submodule keeps the build process simpler than making it an external dependency. (I had a separate fc RPM package in my openSUSE repo, with the idea that it could be used in multiple graphene-based builds. It turned out that each graphene package required its own specific version of fc, which completely defeated the purpose. I'm now building it inline with my current BTS package.)

Not familiar with Meson - what would be the advantage?

`get_next_vote_id` is an implementation thing IMO and does not belong in protocol.

I think `include/config.hpp` contains some protocol-defining things and some implementation-related things. For example, `GRAPHENE_MIN_ACCOUNT_NAME_LENGTH` is used in `protocol/account.cpp` and is the unchangable minimum account name length, whereas `GRAPHENE_MIN_UNDO_HISTORY` is clearly an implementation thing. `protocol/config.hpp` should contain the protocol-related items, `config.hpp` should contain the rest and include `protocol/config.hpp`.

A `pts_address` can be required to authorize a transaction, therefore it should be moved into protocol.

The `db` library is already a separate build. Using the `object_id` type in protocol is arguably legal, IMO - `types.hpp` denotes some database objects as "protocol objects". IDs of these protocol objects are used in operations and they are supposed to be unchangable.
Does it make sense to factor the implementation types out into a separate `include/graphene/chain/types.hpp`?

Currently I don't see any advantages in moving to a different internal database. My experience with chainbase in particular is that its performance goes down the drain when its size exceeds physical RAM, which makes it quite pointless IMO.

**Comment by ryanRfox** [comment]
I assigned @nathanhourt and attempted to pull out a task list from the previous comments, now within the Description. I'll work with @nathanhourt directly about how we can estimate/update as we progress. 

I read some `chainbase` discussion elsewhere, so requesting the results of that be added here.

Thanks!

**Comment by nathanhourt** [comment]
The results are fairly unanimous on chainbase, nobody likes it, so we don't run with it. I'll just make `protocol` dep `db`

I'm also thinking that reorganizing the build system is probably biting off more than I should chew for this particular issue. I maintain that it would be beneficial, but it's not as high a priority as just moving the protocol to its own library.

**Comment by nathanhourt** [comment]
Alright, so here's a question: all of the `XXX_id_type`s are in the `protocol` library, which is necessary because the operations use them... but part of the `XXX_id_type` definitions are the `XXX_object` types ([see here](https://github.com/bitshares/bitshares-core/blob/58c0c494a61c33e5a4ec2312084322f0014b4f5b/libraries/chain/include/graphene/chain/protocol/types.hpp#L186)), which are *not* part of the `protocol` library... but `types.hpp` forward declares them.

Funny part about this is, these types are only used for template magic; they have no impact on the compiled code really, so as far as the `protocol` library is concerned, the forward declarations are sufficient, and it builds fine... But if we're moving `protocol` to its own library, it really belongs in its own namespace as well, and once that happens... it all breaks from the `chain` library's perspective, because now those forward declarations are in the `graphene::protocol` namespace, but the actual types are in the `graphene::chain` namespace... kaboom.

So I can see three alternative solutions to this:
 - Keep protocol library in `graphene::chain` namespace
 - Change `protocol` library's `XXX_object` forward declarations to `graphene::chain` namespace
 - Try to move the protocol-level `XXX_object` types to the `protocol` library

The first two solutions suck as they violate the separation of implementation and protocol... Either way, the `protocol` library now "knows about" the `chain` library, even if only at the namespace level, so it's not unthinkably awful... but it's ugly and kludgy... but it's also easy.

The last solution is the "correct" solution, but IIRC this is a monumental task, if it's possible at all. The `XXX_object`s with protocol-level IDs are riddled with implementation details which are not part of the abstract "Bitshares Protocol," which we never actually defined as distinct from the implementation to that level of detail. I believe the conclusion in the in-office discussions years ago was "all objects are implementation details, even the ones which are implied to exist by the protocol itself."

So I'm requesting input here... Should I just kludge the namespaces? Or am I overestimating the difficulty of moving the protocol-level objects to the `protocol` library?

**Comment by nathanhourt** [comment]
Additional note: I've already made a `chain/types.hpp` and moved declarations regarding the implementation objects to that file, removing them from `protocol/types.hpp`, so the implementation objects wouldn't need to move to `protocol` for solution 3 to work cleanly.

**Comment by pmconrad** [comment]
Hm, tricky.

Side note: I recently learned that the object_id_types raw-serialize only the instance number, not the space nor type id. So technically the database ID (i. e. the tuple `(space,type,instance)` is *not* part of the protocol, only the `instance` is). For the JSON API this is different though.

I can't estimate how much work solution 3 would be. IMO we should use a solution that requires not too much effort. IMO either solution is fine as a first step and can be improved upon later, if we feel the effort is justified.

Actually wrt to the side node above, the ultimate separation would be to use a different ID type at the protocol level that has the same serialization as an object_id_type and can easily be converted from/to database IDs as needed. Not sure if it's worth it.

**Comment by nathanhourt** [comment]
> Actually wrt to the side node above, the ultimate separation would be to use a different ID type at the protocol level that has the same serialization as an object_id_type and can easily be converted from/to database IDs as needed. Not sure if it's worth it.

I like it. I'll see if it plays

**Comment by nathanhourt** [comment]
OK. I think I've got something that works, here. I resolved the ID issue by removing the object type from the ID template, and adding a template mapping that can be defined later which allows the fetching of the object type given the ID type. Now when you call `object_database::get(object_id<>)`, as long as you've already `#include`d the object type's header (which defines the mapping for that object/ID type), you'll get back your expected object type. If you call it without that mapping having been defined (i.e. you didn't include the header), you get back a plain ol' generic `object`.

Check out 6dfe8fe for the gory details.

It builds, it passes tests. I think it's OK... except I got one issue I'm completely confused on. Look at the beginnings of `tests/tests/{call_order,market_rounding,market,settle}_tests.cpp` and you'll see the following lines:
```
#include <graphene/protocol/market.hpp>
#include <graphene/chain/market_object.hpp>
```
The inclusion of `protocol/market.hpp` was not there before, and everything was fine. But now, that header *has* to be included, *before* `chain/market_object.hpp`, or those files fail to build. It took me a few days to realize I could fix those build errors by adding those `#include`s. Can anyone figure out what's going on there? Why does it all break if you swap the order of those two lines (meaning `protocol/market.hpp` gets included later by another header)? It doesn't make any sense to me. @pmconrad, @jmjatlanta Any ideas? This really bugs me :P

**Comment by nathanhourt** [comment]
Oh, you can check out my branch from https://github.com/nathanhourt/bitshares-2/tree/issue-1506

**Comment by pmconrad** [comment]
Wild guess: some headers apply `using` or even `using namespace` to include stuff from other namespaces. This is generally ill-advised and may lead to conflicts and ambiguities.

**Comment by nathanhourt** [comment]
No, I don't think so. The error isn't an ambiguity. As best I can tell, it's unable to find the overload of `fc::raw::pack` specialized for extensions, unless that overload is defined (`#include`d) prior to the definition of the generic overload of `fc::raw::pack`. In other words, it only works if `ext.hpp` is parsed before `fc/io/raw.hpp` is.

Fundamentally, this does not make sense to me. A template isn't evaluated until it's instantiated, so the order in which the templates are declared should not matter, as long as all necessary templates are known by the time things get instantiated. But the compiler seems to be refusing to factor in the specialized `raw::pack` unless it gets declared before the generic `raw::pack`. I almost want to say "compiler bug!" but I'm sure I'd be wrong. I don't understand the rules of function template specialization/overloading that well.

**Comment by nathanhourt** [comment]
Rebase complete -- will preq after merging bitshares/bitshares-fc#123

**Comment by nathanhourt** [comment]
Actually, I'm going to work on this a little more; I've decided it's not ready for integration yet. Making the protocol code a separate library is primarily for the purpose of eliminating its dependency on the chain library. But I also want to eliminate dependency on the db library and even fc. The protocol should stand alone, and not invoke any of our peculiar implementation details, and our database, as well as fc, are most definitely peculiar implementation details.

In short, if I'm going to break out the protocol into its own library, I don't want to half-ass it. I want to do it *right,* making something that's useful to potential future third party implementors.

Towards this end, there are a few steps:
 - [ ] Move all FC references (serialization, exceptions) out of the protocol library and do them outside, in the chain library
 - [ ] Sever dependency on all FC types, or move them into the protocol library (`static_variant` and `unsigned_int` come to mind)
 - [ ] Move `object_id` and `object_id_type` to the protocol library, and reverse the dependency: `db` depends on `protocol`, not vice-versa
 - [ ] Bonus Points: Draw up format specs for protocol objects matching FC's serialization
 - [ ] Major Bonus Points: Write a non-FC, header-only serialization implementation and integrate it into the protocol library

**Comment by nathanhourt** [comment]
> Of the above steps, the first three should be relatively quick and dry.

Lol, yeah, right... This is gonna suck.

**Comment by pmconrad** [comment]
I don't really see the point of moving or re-implementing serialization outside fc. Same for other building blocks like `static_variant`. The only exception to this would be integration of a ready-made drop-in replacement that comes from either `std` or `boost`.

Agree about removing the dependency on db.

**Comment by nathanhourt** [comment]
*sigh* You're right, I get ahead of myself. Replacing fc, even if just in `protocol`, is outside the scope of this issue, and is too big to be worth holding this improvement up for.

But swapping the dependency from `protocol => db` to `db => protocol` is an easy win, so I'll do that.

**Comment by nathanhourt** [comment]
That was an amazingly trivial diff for how much cleaner it makes things conceptually. OK, I'll call this ready to merge once we get the fc change merged (or figure out how to render it irrelevant)

**Comment by abitmore** [comment]
Fixed by #1737.
